include/deal.II/lac/solver_gmres.h

```diff
@@ -1986,16 +2158,112 @@ SolverFGMRES<VectorType>::SolverFGMRES(SolverControl        &cn,
-   concepts::is_linear_operator_on<PreconditionerType, VectorType>))
-void SolverFGMRES<VectorType>::solve(const MatrixType         &A,
-                                     VectorType               &x,
-                                     const VectorType         &b,
-                                     const PreconditionerType &preconditioner)
+   (concepts::is_linear_operator_on<PreconditionerTypes, VectorType> && ...)))
+void SolverMPGMRES<VectorType>::solve(
+  const MatrixType &A,
+  VectorType       &x,
+  const VectorType &b,
+  const PreconditionerTypes &...preconditioners)
 {
-  LogStream::Prefix prefix("FGMRES");
+  LogStream::Prefix prefix("MPGMRES");
+
+  if (additional_data.use_truncated_mpgmres_strategy)
+    SolverMPGMRES<VectorType>::solve_internal(
+      A, x, b, IndexingStrategy::truncated_mpgmres, preconditioners...);
+  else
+    SolverMPGMRES<VectorType>::solve_internal(
+      A, x, b, IndexingStrategy::full_mpgmres, preconditioners...);
+}
+
+
+
+template <typename VectorType>
+DEAL_II_CXX20_REQUIRES(concepts::is_vector_space_vector<VectorType>)
+template <typename MatrixType, typename... PreconditionerTypes>
+void SolverMPGMRES<VectorType>::solve_internal(
+  const MatrixType       &A,
+  VectorType             &x,
+  const VectorType       &b,
+  const IndexingStrategy &indexing_strategy,
+  const PreconditionerTypes &...preconditioners)
+{
+  constexpr std::size_t n_preconditioners = sizeof...(PreconditionerTypes);
+
+  // A lambda for applying the nth preconditioner to a vector src storing
+  // the result in dst:
+
+  const auto apply_nth_preconditioner = [&](unsigned int n,
+                                            auto        &dst,
+                                            const auto  &src) {
+    // We cycle through all preconditioners and call the nth one:
+    std::size_t i = 0;
+
+    [[maybe_unused]] bool preconditioner_called = false;
+
+    const auto call_matching_preconditioner = [&](const auto &preconditioner) {
+      if (i++ == n)
+        {
+          Assert(!preconditioner_called, dealii::ExcInternalError());
+          preconditioner_called = true;
+          preconditioner.vmult(dst, src);
+        }
+    };
+
+    // https://en.cppreference.com/w/cpp/language/fold
+    (call_matching_preconditioner(preconditioners), ...);
+    Assert(preconditioner_called, dealii::ExcInternalError());
+  };
+
+  std::size_t current_index = 0;
+
+  // A lambda that cycles through all preconditioners in sequence while
+  // applying exactly one preconditioner with each function invocation to
+  // the vector src and storing the result in dst:
+
+  const auto preconditioner_vmult = [&](auto &dst, const auto &src) {
+    // We have no preconditioner that we could apply
+    if (n_preconditioners == 0)
+      {
+        dst = src;
+        return;
+      }
+
+    apply_nth_preconditioner(current_index, dst, src);
+    current_index = (current_index + 1) % n_preconditioners;
+  };
```

Comment:
I think this would be easier to read:
```suggestion
  const auto preconditioner_vmult = [&](auto &dst, const auto &src) {
    // We have no preconditioner that we could apply
    if (n_preconditioners == 0)
      dst = src;
    else
      {
        apply_nth_preconditioner(current_index, dst, src);
        current_index = (current_index + 1) % n_preconditioners;
      }
  };
```